### ISSUE TITLE
Assert.IsNotNull(null) in TestCaseSource kills agent in --process=Multiple mode with at least two test assemblies

### ISSUE THREAD

**Issue by MaceWindu** [nunit/nunit#917]
Preconditions:
- console runner called for 2+ assemblies with --process=Multiple switch (process=Separate or only one assembly works perfectly)
- one of assemblies generate failure using Assert.IsNotNull (maybe some other asserts, but at least not Assert.Fail). E.g. using this code:
```
using NUnit.Framework;

public class Test
{
    private static object[] _fail = new object[]
    {
        FailSource
    };

    public static byte[] FailSource
    {
        get
        {
            Assert.IsNotNull(null);
            return null;
        }
    }

    [Test, TestCaseSource("_fail")]
    public void TestMethod(byte[] bytes)
    {
    }
}
```

Results:
after all tests executed, agent for assembly with code above crashes causing runner to crash (see #916)

Code above is just a very simplified version of real code that loads resorce from test assembly and assert loaded blob to be not empty. I plan to create another issue for that, when I will have more data, because resource load works when tests started in different configuration.

**Comment by MaceWindu** [comment]
Whoops, no extra issue required. Looks like assertion doesn't fail (my resources loaded).
```
Assert.IsNotNull(new byte[0]);
```
will also kill agent

**Comment by CharliePoole** [comment]
Are you intending to throw an exception while you are loading the tests?

That's what is happening. NUnit never runs the test because it failed to create it. The entire run should not crash, however I would simply expect those tests to appear as non-runnable.

In any case, using an NUnit Assert outside of your tests isn't going to show up as a failure, merely as an unhandled exception.

**Comment by MaceWindu** [comment]
No, as I said in my comment, agent dies even when assert doesn't generate fail. At least
```
Assert.IsNotNull(new byte[0]);
```
doesn't look for me like a code that must fail.

**Comment by CharliePoole** [comment]
But do you see you are using an Assert outside of any test... before NUnit has even begun to execute tests? That means you are exercising internal infrastructure that has never been initialized. You really don't want an Assert here.

This is a bug because NUnit should protect itself from anything the user code does, even pathological stuff. The fact that it happens because the Assert succeeds, rather than fails, is just a detail.

**Comment by MaceWindu** [comment]
Yep, it's clear. I think I can survive without that assert :)

**Comment by rprouse** [comment]
@MaceWindu, I have a pull request in that fixes this #916, but I could not reproduce with the code you gave above with the current codebase, I needed to cause a stack-overflow to crash the agent. I am thinking this issue is fixed, so I am closing this issue. If you can provide additional repro steps or information, I will be happy to re-open.
